tao_compiler/mlir/disc/tools/disc-transform/TransformOps/TransformOpsExt.cc

```diff
@@ -575,6 +577,309 @@ DiagnosedSilenceableFailure FoldProducerExtractSliceOp::applyToOne(
+    permutation = permutationVec;
+  }
+  build(builder, result, pdl::OperationType::get(builder.getContext()), target,
+        anchor, builder.getI64ArrayAttr(tileLevels),
+        builder.getI64ArrayAttr(tileSizes),
+        builder.getI64ArrayAttr(permutation));
+  if (padded) {
+    result.addAttribute(CacheReadOp::getPaddedAttrName(result.name),
+                        builder.getUnitAttr());
+  }
+}
+
+void CacheReadOp::getEffects(
+    SmallVectorImpl<MemoryEffects::EffectInstance>& effects) {
+  transform::consumesHandle({this->getOperation()->getOperand(0)}, effects);
+  transform::onlyReadsHandle({this->getOperation()->getOperand(1)}, effects);
+  transform::producesHandle(this->getOperation()->getResults(), effects);
+  transform::modifiesPayload(effects);
+}
+
+DiagnosedSilenceableFailure CacheReadOp::apply(
+    transform::TransformResults& results, transform::TransformState& state) {
+  ArrayRef<Operation*> targetOps = state.getPayloadOps(getTarget());
+  if (targetOps.size() != 1)
+    return mlir::emitDefiniteFailure(this->getOperation(),
+                                     "expect only one target op but got ")
+           << targetOps.size();
+
+  Operation* targetOp = targetOps[0];
+  tensor::ExtractSliceOp sliceOp;
+  Optional<Value> paddingValue;
+  if (getPadded()) {
+    auto padOp = dyn_cast<tensor::PadOp>(targetOp);
+    if (!padOp)
+      return mlir::emitDefiniteFailure(targetOp,
+                                       "expect target to be pad op when "
+                                       "`padded` attr is set, but got ")
+             << *targetOp;
+    sliceOp = padOp->getOperand(0).getDefiningOp<tensor::ExtractSliceOp>();
+    if (!sliceOp)
+      return mlir::emitDefiniteFailure(
+          targetOp,
+          "expect the source of pad is a slice op when `padded` attr is set");
+    if (padOp.getBody()->getOperations().size() != 1)
+      return mlir::emitDefiniteFailure(
+          targetOp, "expect the padding region of pad op only has one op\n");
+    paddingValue =
+        cast<tensor::YieldOp>(&padOp.getBody()->getOperations().front())
+            ->getOperand(0);
+  } else {
+    sliceOp = dyn_cast<tensor::ExtractSliceOp>(targetOp);
+    if (!sliceOp)
+      return mlir::emitDefiniteFailure(
+                 targetOp,
+                 "expect target to be extract_slice op when `padded` attr is "
+                 "not set, but got ")
+             << *targetOp;
+  }
+
+  if (!strideAllOnes(sliceOp))
+    return mlir::emitDefiniteFailure(
+        sliceOp.getOperation(),
+        "expect the strides of target slice op are all ones\n");
+
+  // verify the target op have static shape
+  auto targetTy = targetOp->getResult(0).getType().cast<ShapedType>();
+  if (!targetTy.hasStaticShape() || targetTy.getRank() == 0)
+    return mlir::emitDefiniteFailure(
+        targetOp, "expect the targetOp has static shape with rank > 0\n");
+
+  ArrayRef<Operation*> anchorOps = state.getPayloadOps(getAnchor());
```

Comment:
newbie: what is an anchor?

Reply:
anchor op is used to decide where to place the new created `multi_level_pack` op. It's one of the inputs of cache_read op.